lib/pbench/server/api/resources/users_api.py

```diff
@@ -386,6 +407,76 @@ def get(self, username):
             )
             abort(403, message="Authentication token does not belong to the user")
 
+    @Auth.token_auth.login_required(f=Auth().verify_auth())
+    def put(self, username):
+        """
+        PUT request for updating user data.
+        This requires a Pbench auth token in the header field
+
+        This requires a JSON data with required user registration fields that needs an update
+        Example Json:
+        {
+            "first_name": "new_name",
+            "password": "password",
+            ...
+        }
+
+        Required headers include
+
+            Content-Type:   application/json
+            Accept:         application/json
+            Authorization:  Bearer Pbench_auth_token (user received upon login)
+
+        :return: JSON Payload
+            response_object = {
+                    "status": "success"/"fail",
+                    "message": "Successfully registered."/"failure message",
+                    "data": {
+                        "username": <username>,
+                        "first_name": <firstName>,
+                        "last_name": <lastName>,
+                        "registered_on": registered_on,
+                    }
+                }
+        """
+        post_data = request.get_json()
+        try:
+            user, verified = self.auth.verify_user(username)
+        except Exception:
+            self.logger.exception("Exception occurred while verifying the user")
+            abort(500, message="INTERNAL ERROR")
+
+        # TODO: Check if the user has the right privileges
+        if verified or user.is_admin():
+            try:
+                # We will update the user object with the keys and values provided in the request payload.
+                # THe keys need to match with the column names in the user model. However, if any key in
+                # the payload does not match with the column name we just skip that field.
+                user.update(**post_data)
+            except Exception:
+                self.logger.exception("Exception occurred during updating user object")
+                abort(500, message="INTERNAL ERROR")
+        else:
+            self.logger.warning(
+                "Username retrieved from the auth token {} is different from the username provided",
+                self.auth.token_auth.current_user().id,
+            )
+            abort(
+                403, message="Authentication token does not belong to the current user"
+            )
+
+        response_object = {
+            "status": "success",
+            "data": {
+                "username": user.username,
+                "email": user.email,
+                "first_name": user.first_name,
+                "last_name": user.last_name,
+                "registered_on": user.registered_on,
+            },
```

Comment:
I wonder if we should just return the JSON of the user object? If necessary, we can manually sanitize or remove "internal only" properties. This may also tie into the pre-update validation, as there might be some object properties we don't want to allow updating via `PUT` (e.g., creation date).

Reply:
Other than these 5 columns, everything else is just internal properties

Reply:
I added the `get_json()` property on user model so everytime we want to return the user information we call the get_json() and thus it can be managed only in one place.